python/ray/monitor.py

```diff
@@ -235,13 +237,42 @@ def destroy_autoscaler_workers(self):
                 logger.error("Monitor: Cleanup exception. Trying again...")
                 time.sleep(2)
 
+    def _handle_failure(self, error):
+        logger.exception("Error in monitor loop")
+        if self.autoscaler is not None:
+            self.autoscaler.kill_workers()
+            # Take down autoscaler workers if necessary.
+            self.destroy_autoscaler_workers()
+
+        # Something went wrong, so push an error to all current and future
+        # drivers.
+        message = f"The autoscaler failed with the following error:\n{error}"
+        if _internal_kv_initialized():
+            _internal_kv_put(
+                DEBUG_AUTOSCALING_ERROR, message, overwrite=True)
+        redis_client = ray._private.services.create_redis_client(
+            args.redis_address, password=args.redis_password)
+        from ray.utils import push_error_to_driver_through_redis
+        push_error_to_driver_through_redis(
+            redis_client, ray_constants.MONITOR_DIED_ERROR, message)
+
+    def _signal_handler(self, sig, frame):
+        return self._handle_failure(
+            f"Terminated with signal {sig}\n" +
+            "".join(traceback.format_stack(frame)))
+
     def run(self):
+        # Register signal handlers for autoscaler termination.
+        signal.signal(signal.SIGINT, self._signal_handler)
+        signal.signal(signal.SIGTERM, self._signal_handler)
```

Comment:
Signal handlers were easy to add and make it much easier to test the e2e failure handling (i.e. ensuring that the error makes it to the internal kv store and to the PubSub channel).